################################################################################
SHORT DESCRIPTION: 
################################################################################
ACPI support for S3/S4/S5.

################################################################################
LONG DESCRIPTION: 
################################################################################
The values for S3/S4/S5 are defined in a patch to Xen guest ACPI firmware so the
guests knows what values to write. When the guest write the specfic values to
the PM1a_CNT.SLP_TYP, QEMU handles the register write and takes the appropriate
action to shut the guest down.

################################################################################
CHANGELOG 
################################################################################
Port 2.6.2: Ross Philipson, philipsonr@ainfosec.com, 10/13/2016

################################################################################
REMOVAL 
################################################################################

################################################################################
UPSTREAM PLAN 
################################################################################
No.

################################################################################
INTERNAL DEPENDENCIES 
################################################################################
None

################################################################################
PATCHES 
################################################################################
--- a/hw/acpi/core.c
+++ b/hw/acpi/core.c
@@ -20,6 +20,11 @@
  */
 
 #include "qemu/osdep.h"
+#ifdef CONFIG_SYSLOG_LOGGING
+# include "logging.h"
+#else
+# include "stdio.h"
+#endif
 #include "sysemu/sysemu.h"
 #include "hw/hw.h"
 #include "hw/acpi/acpi.h"
@@ -32,6 +37,7 @@
 #include "qemu/error-report.h"
 #include "qemu/module.h"
 #include "qemu/option.h"
+#include "hw/xen/xen.h" /* needed for 'xenstore_update_power' */
 
 struct acpi_table_header {
     uint16_t _length;         /* our length, not actual part of the hdr */
@@ -562,14 +568,17 @@ static void acpi_pm1_cnt_write(ACPIREGS
         switch(sus_typ) {
         case 0: /* soft power off */
             qemu_system_shutdown_request(SHUTDOWN_CAUSE_GUEST_SHUTDOWN);
+            xenstore_update_power(XENSTORE_PM_TYPE_SHUTDOWN);
             break;
         case 1:
             qemu_system_suspend_request();
+            xenstore_update_power(XENSTORE_PM_TYPE_SLEEP);
             break;
         default:
             if (sus_typ == ar->pm1.cnt.s4_val) { /* S4 request */
                 qapi_event_send_suspend_disk();
                 qemu_system_shutdown_request(SHUTDOWN_CAUSE_GUEST_SHUTDOWN);
+                xenstore_update_power(XENSTORE_PM_TYPE_HIBERNATE);
             }
             break;
         }
--- a/hw/acpi/piix4.c
+++ b/hw/acpi/piix4.c
@@ -356,7 +356,15 @@ static void piix4_reset(void *opaque)
         /* Mark SMM as already inited (until KVM supports SMM). */
         pci_conf[0x5B] = 0x02;
     }
-    pm_io_space_update(s);
+
+    /*
+     * OpenXT: Do not update the PM IO space (basically the FADT register
+     * banks) on reset since that will effectively disable them preventing
+     * any further Sx transitions after an S3 is done. The bit in the PCI
+     * config space at 0x80 that enables the PM IO space is only set once
+     * by SeaBIOS at boot time.
+     */
+    /*pm_io_space_update(s);*/
     acpi_pcihp_reset(&s->acpi_pci_hotplug);
 }
 
--- a/hw/i386/xen/xen-hvm.c
+++ b/hw/i386/xen/xen-hvm.c
@@ -1521,6 +1521,15 @@ err:
     exit(1);
 }
 
+static void _xenstore_update_power(bool reboot)
+{
+    fprintf(stderr, "Issued domain %d %s\n", xen_domid,
+          reboot ? "reboot" : "poweroff");
+    if (reboot) {
+      xenstore_update_power(XENSTORE_PM_TYPE_REBOOT);
+    }
+}
+
 void destroy_hvm_domain(bool reboot)
 {
     xc_interface *xc_handle;
@@ -1532,6 +1541,7 @@ void destroy_hvm_domain(bool reboot)
     if (xen_dmod) {
         rc = xendevicemodel_shutdown(xen_dmod, xen_domid, reason);
         if (!rc) {
+            _xenstore_update_power(reboot);
             return;
         }
         if (errno != ENOTTY /* old Xen */) {
@@ -1550,8 +1560,7 @@ void destroy_hvm_domain(bool reboot)
                     "sts %d, %s\n", reboot ? "reboot" : "poweroff",
                     sts, strerror(errno));
         } else {
-            fprintf(stderr, "Issued domain %d %s\n", xen_domid,
-                    reboot ? "reboot" : "poweroff");
+            _xenstore_update_power(reboot);
         }
         xc_interface_close(xc_handle);
     }
--- a/hw/xen/xen-legacy-backend.c
+++ b/hw/xen/xen-legacy-backend.c
@@ -953,3 +953,17 @@ int xenstore_generic_init(void)
 
     return 0;
 }
+
+/* OpenXT: Power Management Notification */
+int xenstore_update_power(enum xenstore_pm_type const type)
+{
+    char *dompath = NULL;
+
+    dompath = xs_get_domain_path(xenstore, xen_domid);
+
+    if (NULL == dompath) {
+        return -1;
+    }
+
+    return xenstore_write_int(dompath, "power-state", type);
+}
--- a/include/hw/xen/xen.h
+++ b/include/hw/xen/xen.h
@@ -46,4 +46,14 @@ void xen_hvm_modified_memory(ram_addr_t
 
 void xen_register_framebuffer(struct MemoryRegion *mr);
 
+/* Power Management */
+enum xenstore_pm_type {
+    XENSTORE_PM_TYPE_REBOOT = 0,
+    XENSTORE_PM_TYPE_SLEEP = 3,
+    XENSTORE_PM_TYPE_HIBERNATE = 4,
+    XENSTORE_PM_TYPE_SHUTDOWN = 5
+};
+
+int xenstore_update_power(enum xenstore_pm_type const type);
+
 #endif /* QEMU_HW_XEN_H */
